File: python/copass-agent-router/src/copass_agent_router/sse.py
"""Minimal SSE parser for Copass's agent-run endpoint.

Translates raw SSE frames into neutral :class:`AgentEvent` values from
``copass-core-agents``. Handles CRLF/LF line endings, multi-line
``data:`` fields, and comment/id/retry skipping.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import AsyncIterator, Optional

import httpx

from copass_core_agents.events import (
    AgentEvent,
    AgentFinish,
    AgentTextDelta,
    AgentToolCall,
    AgentToolResult,
)


@dataclass(frozen=True)
class RawSseFrame:
    event: str
    data: str


def _parse_block(block: str) -> Optional[RawSseFrame]:
    event = "message"
    data_lines: list[str] = []
    for raw in block.split("\n"):
        line = raw.rstrip("\r")
        if not line or line.startswith(":"):
            continue
        colon = line.find(":")
        if colon < 0:
            continue
        field = line[:colon]
        value = line[colon + 1 :]
        if value.startswith(" "):
            value = value[1:]
        if field == "event":
            event = value
        elif field == "data":
            data_lines.append(value)
    if not data_lines:
        return None
    return RawSseFrame(event=event, data="\n".join(data_lines))


async def iterate_sse_frames(response: httpx.Response) -> AsyncIterator[RawSseFrame]:
    """Async-iterate SSE frames off an ``httpx`` streaming Response.

    Caller is expected to have opened the response with ``stream=True``
    (i.e. via ``client.stream(...)``). We don't read the full body in
    one shot — each frame is yielded as it arrives.
    """
    buffer = ""
    async for chunk in response.aiter_text():
        buffer += chunk
        buffer = buffer.replace("\r\n", "\n")
        while True:
            sep = buffer.find("\n\n")
            if sep < 0:
                break
            block = buffer[:sep]
            buffer = buffer[sep + 2 :]
            frame = _parse_block(block)
            if frame is not None:
                yield frame
    tail = buffer.strip()
    if tail:
        frame = _parse_block(tail)
        if frame is not None:
            yield frame


def frame_to_agent_event(frame: RawSseFrame) -> Optional[AgentEvent]:
    """Translate a Copass SSE frame into a neutral :class:`AgentEvent`.

    Returns ``None`` for unrecognized event names or malformed JSON.
    """
    try:
        payload = json.loads(frame.data)
    except (ValueError, TypeError):
        return None
    if not isinstance(payload, dict):
        return None
    if frame.event == "agent_text_delta":
        return AgentTextDelta(text=str(payload.get("text", "")))
    if frame.event == "agent_tool_call":
        return AgentToolCall(
            call_id=str(payload.get("call_id", "")),
            name=str(payload.get("name", "")),
            arguments=dict(payload.get("arguments") or {}),
        )
    if frame.event == "agent_tool_result":
        return AgentToolResult(
            call_id=str(payload.get("call_id", "")),
            name=str(payload.get("name", "")),
            result=dict(payload.get("result") or {}),
            error=(str(payload["error"]) if payload.get("error") else None),
        )
    if frame.event == "agent_finish":
        return AgentFinish(
            stop_reason=str(payload.get("stop_reason", "unknown")),
            session_id=(payload.get("session_id") or None),
            usage=dict(payload.get("usage") or {}),
        )
    return None


__all__ = ["RawSseFrame", "iterate_sse_frames", "frame_to_agent_event"]

File: typescript/packages/agent-router/src/sse.ts
/**
 * Minimal SSE (Server-Sent Events) parser for `text/event-stream` responses.
 *
 * Handles:
 * - CRLF and LF line endings
 * - Multi-line `data:` fields (concatenated per spec)
 * - `event:` field for message type
 * - Blank line as frame separator
 * - Ignores `:` comments, `id:`, `retry:`
 *
 * Copass's agent router emits frames tagged as `agent_text_delta`,
 * `agent_tool_call`, `agent_tool_result`, `agent_finish`, `agent_error`.
 */

import type {
  AgentEvent,
  AgentErrorEvent,
  AgentFinish,
  AgentTextDelta,
  AgentToolCall,
  AgentToolResult,
} from './events.js';

export interface RawSseFrame {
  event: string;
  data: string;
}

/** Parse one SSE frame block (events separated by blank lines). */
function parseBlock(block: string): RawSseFrame | null {
  let eventName = 'message';
  const dataLines: string[] = [];
  for (const rawLine of block.split('\n')) {
    const line = rawLine.replace(/\r$/, '');
    if (!line || line.startsWith(':')) continue;
    const colon = line.indexOf(':');
    if (colon < 0) continue;
    const field = line.slice(0, colon);
    // Strip one leading space after the colon per spec
    const value = line.slice(colon + 1).replace(/^ /, '');
    if (field === 'event') eventName = value;
    else if (field === 'data') dataLines.push(value);
  }
  if (dataLines.length === 0) return null;
  return { event: eventName, data: dataLines.join('\n') };
}

/** Async-iterate over SSE frames from a Fetch Response body. */
export async function* iterateSseFrames(
  response: Response,
): AsyncIterableIterator<RawSseFrame> {
  if (!response.body) return;
  const reader = response.body.getReader();
  const decoder = new TextDecoder('utf-8');
  let buffer = '';
  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      // Normalize CRLF → LF; split on blank line
      buffer = buffer.replace(/\r\n/g, '\n');
      let idx: number;
      while ((idx = buffer.indexOf('\n\n')) >= 0) {
        const block = buffer.slice(0, idx);
        buffer = buffer.slice(idx + 2);
        const frame = parseBlock(block);
        if (frame) yield frame;
      }
    }
    // Flush trailing block if any
    if (buffer.trim()) {
      const frame = parseBlock(buffer);
      if (frame) yield frame;
    }
  } finally {
    try {
      reader.releaseLock();
    } catch {
      /* ignore */
    }
  }
}

/** Translate a raw Copass SSE frame into the neutral `AgentEvent` union. */
export function frameToAgentEvent(frame: RawSseFrame): AgentEvent | null {
  let payload: Record<string, unknown>;
  try {
    payload = JSON.parse(frame.data);
  } catch {
    return null;
  }
  switch (frame.event) {
    case 'agent_text_delta':
      return { type: 'text', text: String(payload.text ?? '') } as AgentTextDelta;
    case 'agent_tool_call':
      return {
        type: 'tool_call',
        call_id: String(payload.call_id ?? ''),
        name: String(payload.name ?? ''),
        arguments: (payload.arguments as Record<string, unknown>) ?? {},
      } as AgentToolCall;
    case 'agent_tool_result':
      return {
        type: 'tool_result',
        call_id: String(payload.call_id ?? ''),
        name: String(payload.name ?? ''),
        result: (payload.result as Record<string, unknown>) ?? {},
        error: (payload.error as string | null | undefined) ?? null,
      } as AgentToolResult;
    case 'agent_finish':
      return {
        type: 'finish',
        stop_reason: String(payload.stop_reason ?? 'unknown'),
        session_id: (payload.session_id as string | null | undefined) ?? null,
        usage: (payload.usage as Record<string, unknown>) ?? {},
      } as AgentFinish;
    case 'agent_error':
      return {
        type: 'error',
        message: String(payload.message ?? ''),
        errorType: String(payload.type ?? 'Error'),
      } as AgentErrorEvent;
    default:
      return null;
  }
}
